File: common/src/main/java/com/wdz/common/view/CommonTitleBar.java
package com.wdz.common.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.constraintlayout.widget.ConstraintLayout;

import com.wdz.common.R;
import com.wdz.common.util.ScreenUtils;


/**
 * 通用titleBar，高度自适应，宽度match
 * @author wdz
 */
public class CommonTitleBar extends ConstraintLayout {
    private OnClickListener onClickListener;
    private TextView tvTitleName;
    private ImageView backImg;
    private ImageView closeImg;
    private ConstraintLayout titleBarBg;
    private TypedArray typedArray;
    private ImageView leftImg;
    private ImageView rightImg;
    private TextView textRight;

    public CommonTitleBar(Context context) {
        this(context,null);
    }

    public CommonTitleBar(Context context, AttributeSet attrs) {
        this(context,attrs,0);
    }

    public CommonTitleBar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView(context,attrs);

    }

    /**
     * 加载视图
     * @param context
     * @param attrs
     */
    private void initView(Context context, AttributeSet attrs) {
        initUi(context);
        initStyle(context,attrs);


    }

    /**
     * 初始化ui
     * @param context
     */
    private void initUi(Context context) {
        int statusBarHeight = ScreenUtils.getStatusBarHeight(context);
        View view = View.inflate(context, R.layout.title_bar_common, this);
        titleBarBg = findViewById(R.id.cl_root);
        tvTitleName = view.findViewById(R.id.tv_title);
        textRight = view.findViewById(R.id.text_right);
        backImg = view.findViewById(R.id.iv_nav_back);
        leftImg = view.findViewById(R.id.iv_left);
        closeImg = view.findViewById(R.id.iv_nav_close);
        rightImg = view.findViewById(R.id.iv_right);
        LayoutParams params = (LayoutParams) titleBarBg.getLayoutParams();
        params.height = dp2px(context,44) + statusBarHeight;
        params.width = LayoutParams.MATCH_PARENT;
        titleBarBg.setLayoutParams(params);
        backImg.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (onClickListener!=null){
                    onClickListener.onBackPress();
                }
            }
        });
        closeImg.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (onClickListener!=null){
                    onClickListener.onClosePress();
                }
            }
        });
        rightImg.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (onClickListener!=null){
                    onClickListener.onSettingPress();
                }
            }
        });
        textRight.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (onClickListener!=null){
                    onClickListener.onClickRightText();
                }
            }
        });
    }


    /**
     * 加载标题style
     * @param context
     * @param attrs
     */
    private void initStyle(Context context,AttributeSet attrs) {
        typedArray = context.obtainStyledAttributes(attrs, R.styleable.CommonTitleBar);
        //设置背景颜色
        if (typedArray.hasValue(R.styleable.CommonTitleBar_background_color)){
            titleBarBg.setBackgroundColor(typedArray.getColor(R.styleable.CommonTitleBar_background_color,0xFFFFFF));
        }
        if (typedArray.hasValue(R.styleable.CommonTitleBar_title)){
            setTitle();
        }
        //最左侧图标 -- back
        if (typedArray.hasValue(R.styleable.CommonTitleBar_icon_back)){
            backImg.setImageResource(typedArray.getResourceId(R.styleable.CommonTitleBar_icon_back,R.mipmap.ic_return));
        }
        //左二图标 -- 预留
        if (typedArray.hasValue(R.styleable.CommonTitleBar_icon_left)){
            leftImg.setImageResource(typedArray.getResourceId(R.styleable.CommonTitleBar_icon_left,R.mipmap.ic_return));
        }
        //最右侧图标 -- close
        if (typedArray.hasValue(R.styleable.CommonTitleBar_icon_right)){
            rightImg.setImageResource(typedArray.getResourceId(R.styleable.CommonTitleBar_icon_right,R.mipmap.ic_add));
        }
        //右二图标 -- 一般为设置图标...
        if (typedArray.hasValue(R.styleable.CommonTitleBar_icon_close)){
            closeImg.setImageResource(typedArray.getResourceId(R.styleable.CommonTitleBar_icon_close,R.mipmap.ic_add));
        }

        //右侧文字
        if (typedArray.hasValue(R.styleable.CommonTitleBar_text_right)){
            textRight.setText(typedArray.getString(R.styleable.CommonTitleBar_text_right));
        }
    }

    /**
     * 设置标题相关文字、颜色、大小、样式
     */
    private void setTitle() {
        tvTitleName.setText(typedArray.getString(R.styleable.CommonTitleBar_title));
        if (typedArray.hasValue(R.styleable.CommonTitleBar_title_size)){
            tvTitleName.setTextSize(typedArray.getDimensionPixelSize(R.styleable.CommonTitleBar_title_size,sp2px(getContext(),18)));
        }
        if (typedArray.hasValue(R.styleable.CommonTitleBar_title_color)){
            tvTitleName.setTextColor(typedArray.getColor(R.styleable.CommonTitleBar_title_color,0x333333));
        }
        if (typedArray.hasValue(R.styleable.CommonTitleBar_title_style)){
            //只有常规与加粗两种
            if (typedArray.getInt(R.styleable.CommonTitleBar_title_style,0) == 0){
                tvTitleName.setTypeface(Typeface.DEFAULT);
            }
            else if (typedArray.getInt(R.styleable.CommonTitleBar_title_style,0) == 1){
                tvTitleName.setTypeface(Typeface.DEFAULT_BOLD);
            }

        }

    }

    /**
     * 获取bar高度
     * @return
     */
    public int getBarHeight(){
        return titleBarBg.getHeight();
    }

    /**
     * 设置标题名称
     * @param titleName
     */
    public void setTitleName(String titleName){
        tvTitleName.setText(titleName);
    }



    /**
     * 设置TitleBar的点击事件
     * @param onClickListener
     */
    public void setOnTitleBarClickListener(OnClickListener onClickListener){
        this.onClickListener = onClickListener;
    }

    /**
     * 点击监听
     */
    public interface OnClickListener{
        /**
         * 点击返回按钮
         */
        void onBackPress();
        /**
         * 点击关闭按钮
         */
        void onClosePress();
        /**
         * 点击设置按钮
         */
        void onSettingPress();
        /**
         * 点击右侧文本
         */
        void onClickRightText();
    }



    /**
     * sp -> px
     *
     * @param context
     * @param spValue
     * @return
     */
    public int sp2px(Context context, float spValue) {
        final float fontScale = context.getResources().getDisplayMetrics().scaledDensity;
        return (int) (spValue * fontScale + 0.5f);
    }

    /**
     * dp->px
     * @param context  context
     * @param dipValue dp值
     * @return px值
     */
    public static int dp2px(Context context, float dipValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dipValue * scale + 0.5f);
    }
}

File: kt_common/src/main/java/com/wdz/ktcommon/view/CommonTitleBar.kt
package com.wdz.ktcommon.view

import android.content.Context
import android.content.res.TypedArray
import android.graphics.Typeface
import android.util.AttributeSet
import android.view.View

import androidx.constraintlayout.widget.ConstraintLayout
import com.wdz.ktcommon.R
import com.wdz.ktcommon.utils.getStatusBarHeight
import kotlinx.android.synthetic.main.title_bar_common.view.*


/**
 * 通用titleBar，高度自适应，宽度match
 * @author wdz
 */
class CommonTitleBar : ConstraintLayout {
    private var onClickListener: OnClickListener? = null
    private var typedArray: TypedArray? = null

    constructor(context: Context):super(context){
        initView(context, null)
    }

    constructor(context: Context,attrs: AttributeSet):super(context,attrs){
        initView(context, attrs)
    }

    constructor(context: Context,attrs: AttributeSet,defStyleAttr:Int):super(context,attrs,defStyleAttr){
        initView(context, attrs)
    }


    /**
     * 加载视图
     * @param context
     * @param attrs
     */
    private fun initView(context: Context, attrs: AttributeSet?) {
        initUi(context)
        initStyle(context, attrs)
    }

    /**
     * 初始化ui
     * @param context
     */
    private fun initUi(context: Context) {
        View.inflate(context, R.layout.title_bar_common, this)

        val params = cl_root.layoutParams
        params.height = dp2px(context, 44f) + getStatusBarHeight(context)
        params.width = LayoutParams.MATCH_PARENT
        cl_root.layoutParams = params
        iv_nav_back.setOnClickListener(View.OnClickListener {
            onClickListener?.onBackPress()
        })
        iv_nav_close.setOnClickListener(View.OnClickListener {
            onClickListener?.onClosePress()
        })
        iv_right.setOnClickListener(View.OnClickListener {
            onClickListener?.onSettingPress()
        })
        text_right.setOnClickListener(View.OnClickListener {
            onClickListener?.onClickRightText()
        })
    }

    /**
     * 加载标题style
     * @param context
     * @param attrs
     */
    private fun initStyle(
        context: Context,
        attrs: AttributeSet?
    ) {
        typedArray = context.obtainStyledAttributes(attrs, R.styleable.CommonTitleBar)

        typedArray?.run {
            //设置背景颜色
            if (hasValue(R.styleable.CommonTitleBar_background_color)){
                cl_root.setBackgroundColor(getColor(
                    R.styleable.CommonTitleBar_background_color,
                        0xFFFFFF))
            }

            if (hasValue(R.styleable.CommonTitleBar_title)) {
                setTitle()
            }

            //最左侧图标 -- back
            if (hasValue(R.styleable.CommonTitleBar_icon_back)) {
                iv_nav_back.setImageResource(getResourceId(
                    R.styleable.CommonTitleBar_icon_back, R.mipmap.ic_return))
            }

            //左二图标 -- 预留
            if (hasValue(R.styleable.CommonTitleBar_icon_left)) {
                iv_left.setImageResource(
                    getResourceId(R.styleable.CommonTitleBar_icon_left, R.mipmap.ic_return))
            }

            //最右侧图标 -- close
            if (hasValue(R.styleable.CommonTitleBar_icon_right)) {
                iv_right.setImageResource(
                    getResourceId(R.styleable.CommonTitleBar_icon_right, R.mipmap.ic_add))
            }

            //右二图标 -- 一般为设置图标...
            if (hasValue(R.styleable.CommonTitleBar_icon_close)) {
                iv_nav_close.setImageResource(
                    getResourceId(R.styleable.CommonTitleBar_icon_close, R.mipmap.ic_add)
                )
            }

            //右侧文字
            if (hasValue(R.styleable.CommonTitleBar_text_right)) {
                text_right.text = getString(R.styleable.CommonTitleBar_text_right)
            }
        }

    }

    /**
     * 设置标题相关文字、颜色、大小、样式
     */
    private fun setTitle() {
        typedArray?.run {
            tv_title.text = getString(R.styleable.CommonTitleBar_title)
            if (hasValue(R.styleable.CommonTitleBar_title_size)) {
                tv_title.textSize = getDimensionPixelSize(
                    R.styleable.CommonTitleBar_title_size,
                    sp2px(context, 18f)
                ).toFloat()
            }
            if (hasValue(R.styleable.CommonTitleBar_title_color)) {
                tv_title.setTextColor(
                    getColor(
                        R.styleable.CommonTitleBar_title_color,
                        0x333333
                    )
                )
            }
            if (hasValue(R.styleable.CommonTitleBar_title_style)) {
                //只有常规与加粗两种
                if (getInt(R.styleable.CommonTitleBar_title_style, 0) == 0) {
                    tv_title.typeface = Typeface.DEFAULT
                } else if (getInt(R.styleable.CommonTitleBar_title_style, 0) == 1) {
                    tv_title!!.typeface = Typeface.DEFAULT_BOLD
                }
            }
        }

    }

    /**
     * 获取bar高度
     * @return
     */
    val barHeight: Int
        get() = cl_root.height

    /**
     * 设置标题名称
     * @param titleName
     */
    fun setTitleName(titleName: String?) {
        tv_title.text = titleName
    }

    /**
     * 设置TitleBar的点击事件
     * @param onClickListener
     */
    fun setOnTitleBarClickListener(onClickListener: OnClickListener?) {
        this.onClickListener = onClickListener
    }

    /**
     * 点击监听
     */
    interface OnClickListener {
        /**
         * 点击返回按钮
         */
        fun onBackPress()

        /**
         * 点击关闭按钮
         */
        fun onClosePress()

        /**
         * 点击设置按钮
         */
        fun onSettingPress()

        /**
         * 点击右侧文本
         */
        fun onClickRightText()
    }

    /**
     * sp -> px
     *
     * @param context
     * @param spValue
     * @return
     */
    private fun sp2px(context: Context, spValue: Float): Int {
        val fontScale = context.resources.displayMetrics.scaledDensity
        return (spValue * fontScale + 0.5f).toInt()
    }

    /**
     * dp->px
     * @param context  context
     * @param dipValue dp值
     * @return px值
     */
    private fun dp2px(context: Context, dipValue: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (dipValue * scale + 0.5f).toInt()
    }


}